samples/bluetooth/alexa_gadget/src/gadgets/CMakeLists.txt

```diff
@@ -130,29 +130,30 @@ foreach(GROUP BASE NOTIFICATIONS ALERTS STATELISTENER SPEECHDATA MUSICDATA CUSTO
         get_filename_component(FILE_BASE ${file} NAME_WE)
         list(APPEND PROTOC_INC_DIRS "-I.")
         list(APPEND PROTOC_INC_DIRS "-I${FILE_PATH}")
-        list(APPEND ${GROUP}_PROTO_BYPRODUCTS "${PROTO_COMPILATION_DIR}/${FILE_BASE}.pb.c")
+        list(APPEND ${GROUP}_PROTO_OUTPUTS "${PROTO_COMPILATION_DIR}/${FILE_BASE}.pb.c")
+        list(APPEND ${GROUP}_PROTO_DEPENDENCIES "${file}")
     endforeach()
 
-    # Populate list of compilation commands for protoc compiler.
-    unset(PROTOC_CMD_LIST)
+    # Each .pb.c file to be compiled by protoc gets its own CMake custom command,
+    # with dependencies to all other .proto files in its group.
     foreach(file ${${GROUP}_PROTO_FILES})
         get_filename_component(FILE_PATH ${file} DIRECTORY)
         get_filename_component(FILE_NAME ${file} NAME)
+        get_filename_component(FILE_BASE ${file} NAME_WE)
+
         set(PROTOC_ARGS "-I."
                         "${PROTOC_INC_DIRS}"
                         "--nanopb_out=${PROTO_COMPILATION_DIR}"
                         "${FILE_NAME}"
         )
-        list(APPEND PROTOC_CMD_LIST
-                    COMMAND ${CMAKE_COMMAND} -E chdir ${FILE_PATH}
-                    ${PROTOC_BIN} ${PROTOC_ARGS})
-    endforeach()
 
-    add_custom_target(
-        protobuf_generation_${GROUP}
-        ${PROTOC_CMD_LIST}
-        BYPRODUCTS ${${GROUP}_PROTO_BYPRODUCTS}
-    )
+        add_custom_command(
+          OUTPUT "${PROTO_COMPILATION_DIR}/${FILE_BASE}.pb.c"
+          COMMAND ${CMAKE_COMMAND} -E chdir ${FILE_PATH} ${PROTOC_BIN} ${PROTOC_ARGS}
+          MAIN_DEPENDENCY ${file}
+          DEPENDS ${${GROUP}_PROTO_DEPENDENCIES}
+        )
```

Comment:
Use `WORKING_DIRECTORY ${FILE_PATH}` instead of cmake -E chdir?

Reply:
> 
> 
> Use `WORKING_DIRECTORY ${FILE_PATH}` instead of cmake -E chdir?

I recall @tejlmand recommended using `cmake -E` instead of originally `cd working_dir && binary --parameters`. I do like the improved readability of the suggested `WORKING_DIRECTORY ${FILE_PATH}`, but want to confirm it doesn't have any side effects

Reply:
`cmake -E chdir` is preferred over `cd` since cmake is platform independent.
But in order to run the custom command in a different folder it is normal to use WORKING_DIRECTORY instead.

Reply:
The `cmake -E chdir` was because of the `&&` amd the use of `list(append ...)`.

Using of `list(APPEND PROTOC_CMD_LIST ...)` then each `COMMAND` could potentially be using it's own working directory, and the use of `cmake -E chdir` would be the only way to ensure that.

If there is only one command, like in the new proposal, then `WORKING_DIRECTORY` should be used.
(or when all commands are to be executed in same work dir)